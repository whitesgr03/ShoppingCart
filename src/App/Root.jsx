import { useState, useEffect } from "react";
import { Outlet, useLocation, useNavigate } from "react-router-dom";


import Header from "./Header";
import Footer from "./Footer";
import Modal from "../components/modals/Modal";

import RootProvider from "./RootContext";
import fetchBackgroundImageUrl from "../utils/fetchBackgroundImageUrl";

const Root = () => {
	const [isLoading, setIsLoading] = useState(false);
	const [imageUrl, setImageUrl] = useState("");
	const [backgroundImage, setBackgroundImage] = useState({
		home: null,
		contact: null,
	});
	const location = useLocation();
	const navigate = useNavigate();

	const page =
		location.pathname.slice(1) === "" ? "home" : location.pathname.slice(1);

	const imageLoad = () => {
		setBackgroundImage({
			...backgroundImage,
			[page]: `url(${imageUrl})`,
		});

		setImageUrl("");
		setIsLoading(false);
	};

	useEffect(() => {
		const handleFetch = async () => {
			setIsLoading(true);
			try {
				setImageUrl(await fetchBackgroundImageUrl(page));
			} catch (error) {
				console.log({
					message: "fetch image does not exist",
					state: error,
				});
				navigate("/error");
				setIsLoading(false);
			}
		};

		backgroundImage[page] === null && handleFetch();
	}, [page, backgroundImage, navigate]);

	return (
		<>
			<RootProvider>
				<Header />
				<div data-testid="content" className={"content"}>
					{isLoading ? (
						<div data-testid="loading" className="loading">
							<img
								onLoad={imageLoad}
								src={imageUrl}
								alt="none"
								hidden
							/>
							<Icon path={mdiLoading} spin={1} size={3} />
							Loading...
						</div>
					) : (
						<Outlet
							context={{
								products,
								backgroundImage,
							}}
						/>
					)}
					<Footer />
				</div>
				<Modal />
			</RootProvider>
		</>
	);
};

export { Root as default };
