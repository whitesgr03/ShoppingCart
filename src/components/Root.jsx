import { useState } from "react";
import { Outlet } from "react-router-dom";

import Header from "./Header";
import Footer from "./Footer";
import Modal from "./Modal";

const Root = () => {
	const [cartList, setCartList] = useState([]);
	const [showModal, setShowModal] = useState(null);
	const [latestItem, setLatestItem] = useState(null);
	

	const handleRemoveItem = id => {
		setCartList(cartList.filter(item => item.id !== id));
	};

	const handleEditItem = (id, quantity) => {
		setCartList(
			cartList.map(item =>
				item.id === id
					? {
							...item,
							quantity,
					  }
					: item
			)
		);
	};

	const handleAddItem = product => {
		setCartList(
			cartList.find(item => item.id === product.id)
				? cartList.map(item =>
						item.id === product.id
							? {
									...item,
									quantity: item.quantity + product.quantity,
							  }
							: item
				  )
				: [...cartList, product]
		);

		setLatestItem(product);
	};

	const handleToggleModal = value => setShowModal(value);

	return (
		<>
			<Header cartList={cartList} onToggleModal={handleToggleModal} />
			<div data-testid="content" className={"content"}>
				<Outlet
					context={{
						onAddItem: handleAddItem,
						onToggleModal: handleToggleModal,
					}}
				/>
				<Footer />
			</div>
			<Modal
				cartList={cartList}
				showModal={showModal}
				onToggleModal={handleToggleModal}
			/>
		</>
	);
};

export { Root as default };
