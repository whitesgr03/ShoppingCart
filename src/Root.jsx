import App from "./components/App";
import Home from "./components/Home";
import ShopPage from "./components/products/ShopPage";
import ProductList from "./components/products/ProductList";
import SingleProductPage from "./components/products/SingleProductPage";
import Contact from "./components/Contact";
import AuthGuard from "./components/AuthGuard";
import NotFound from "./components/NotFound";

import { createBrowserRouter, RouterProvider } from "react-router-dom";

const Root = () => {
	const router = createBrowserRouter([
		{
			path: "/",
			element: <App />,
			errorElement: <NotFound />,
				},
			],
		},
	]);
	return <RouterProvider router={router} />;
};

export default Root;
