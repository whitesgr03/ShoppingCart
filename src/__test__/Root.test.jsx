import { render } from "@testing-library/react";

import { MemoryRouter } from "react-router-dom";

import Root from "../components/Root";

describe("Renders Root Component", () => {
	it("Should return Root DOM", () => {
		const { container } = render(
			<MemoryRouter initialEntries={["/"]}>
				<Root />
			</MemoryRouter>
		);

		const actual = container;

		expect(actual).toMatchSnapshot();
	});
});
