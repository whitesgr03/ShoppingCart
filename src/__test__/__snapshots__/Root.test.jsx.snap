// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Root Component Should return Root DOM 1`] = `
<div>
  <div
    class="sidebar"
  >
    <div
      class="icons"
    >
      <button
        type="button"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,4A4,4 0 0,1 16,8A4,4 0 0,1 12,12A4,4 0 0,1 8,8A4,4 0 0,1 12,4M12,14C16.42,14 20,15.79 20,18V20H4V18C4,15.79 7.58,14 12,14Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
      <button
        type="button"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M5,3C3.89,3 3,3.89 3,5V19C3,20.11 3.89,21 5,21H19C20.11,21 21,20.11 21,19V5C21,3.89 20.11,3 19,3H5M5,5H19V19H5V5M7,7V9H17V7H7M7,11V13H17V11H7M7,15V17H14V15H7Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
      <button
        type="button"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 20C19 21.11 18.11 22 17 22C15.89 22 15 21.1 15 20C15 18.89 15.89 18 17 18C18.11 18 19 18.9 19 20M7 18C5.89 18 5 18.89 5 20C5 21.1 5.89 22 7 22C8.11 22 9 21.11 9 20S8.11 18 7 18M7.2 14.63L7.17 14.75C7.17 14.89 7.28 15 7.42 15H19V17H7C5.89 17 5 16.1 5 15C5 14.65 5.09 14.32 5.24 14.04L6.6 11.59L3 4H1V2H4.27L5.21 4H20C20.55 4 21 4.45 21 5C21 5.17 20.95 5.34 20.88 5.5L17.3 11.97C16.96 12.58 16.3 13 15.55 13H8.1L7.2 14.63M8.5 11H10V9H7.56L8.5 11M11 9V11H14V9H11M14 8V6H11V8H14M17.11 9H15V11H16L17.11 9M18.78 6H15V8H17.67L18.78 6M6.14 6L7.08 8H10V6H6.14Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
    </div>
    <h1
      class="title"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        GentSkin
      </a>
    </h1>
    <ul
      class="pages"
    >
      <li>
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          class=""
          href="/products"
        >
          Shop
        </a>
      </li>
      <li>
        <a
          class=""
          href="/contact"
        >
          Contact
        </a>
      </li>
    </ul>
  </div>
  <div
    class="content"
  >
    <footer>
      <a
        href="https://github.com/whitesgr03/ShoppingCart"
        rel="noreferrer"
        target="_blank"
      >
        Â© 2023 - GentSkin
      </a>
      <a
        href="https://github.com/whitesgr03/ShoppingCart"
        rel="noreferrer"
        target="_blank"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,2A10,10 0 0,0 2,12C2,16.42 4.87,20.17 8.84,21.5C9.34,21.58 9.5,21.27 9.5,21C9.5,20.77 9.5,20.14 9.5,19.31C6.73,19.91 6.14,17.97 6.14,17.97C5.68,16.81 5.03,16.5 5.03,16.5C4.12,15.88 5.1,15.9 5.1,15.9C6.1,15.97 6.63,16.93 6.63,16.93C7.5,18.45 8.97,18 9.54,17.76C9.63,17.11 9.89,16.67 10.17,16.42C7.95,16.17 5.62,15.31 5.62,11.5C5.62,10.39 6,9.5 6.65,8.79C6.55,8.54 6.2,7.5 6.75,6.15C6.75,6.15 7.59,5.88 9.5,7.17C10.29,6.95 11.15,6.84 12,6.84C12.85,6.84 13.71,6.95 14.5,7.17C16.41,5.88 17.25,6.15 17.25,6.15C17.8,7.5 17.45,8.54 17.35,8.79C18,9.5 18.38,10.39 18.38,11.5C18.38,15.32 16.04,16.16 13.81,16.41C14.17,16.72 14.5,17.33 14.5,18.26C14.5,19.6 14.5,20.68 14.5,21C14.5,21.27 14.66,21.59 15.17,21.5C19.14,20.16 22,16.42 22,12A10,10 0 0,0 12,2Z"
            style="fill: currentColor;"
          />
        </svg>
      </a>
    </footer>
  </div>
</div>
`;
